sklearn/inspection/_base.py

```diff
@@ -0,0 +1,365 @@
+            axis = uniques
+        else:
+            # create axis based on percentiles and grid resolution
+            emp_percentiles = mquantiles(
+                _safe_indexing(X, feature, axis=1), prob=percentiles, axis=0
+            )
+            if np.allclose(emp_percentiles[0], emp_percentiles[1]):
+                raise ValueError(
+                    'percentiles are too close to each other, '
+                    'unable to build the grid. Please choose percentiles '
+                    'that are further apart.')
+            axis = np.linspace(emp_percentiles[0],
+                               emp_percentiles[1],
+                               num=grid_resolution, endpoint=True)
+        values.append(axis)
+
+    return cartesian(values), values
+
+
+def _get_predictions(est, grid, features_indices, X, response_method):
+    predictions = []
+
+    # define the prediction_method (predict, predict_proba, decision_function).
+    if is_regressor(est):
+        prediction_method = est.predict
+    else:
+        predict_proba = getattr(est, 'predict_proba', None)
+        decision_function = getattr(est, 'decision_function', None)
+        if response_method == 'auto':
+            # try predict_proba, then decision_function if it doesn't exist
+            prediction_method = predict_proba or decision_function
+        else:
+            prediction_method = (predict_proba if response_method ==
+                                 'predict_proba' else decision_function)
+        if prediction_method is None:
+            if response_method == 'auto':
+                raise ValueError(
+                    'The estimator has no predict_proba and no '
+                    'decision_function method.'
+                )
+            elif response_method == 'predict_proba':
+                raise ValueError('The estimator has no predict_proba method.')
+            else:
+                raise ValueError(
+                    'The estimator has no decision_function method.')
+
+    for new_values in grid:
+        X_eval = X.copy()
+        for i, variable in enumerate(features_indices):
+            if hasattr(X_eval, 'iloc'):
+                X_eval.iloc[:, variable] = new_values[i]
+            else:
+                X_eval[:, variable] = new_values[i]
+
+        try:
+            # Note: predictions is of shape
+            # (n_points,) for non-multioutput regressors
+            # (n_points, n_tasks) for multioutput regressors
+            # (n_points, 1) for the regressors in cross_decomposition (I think)
+            # (n_points, 2) for binary classification
+            # (n_points, n_classes) for multiclass classification
+            predictions.append(prediction_method(X_eval))
+        except NotFittedError:
+            raise ValueError(
+                "'estimator' parameter must be a fitted estimator")
+
+    return predictions
+
+
+def _validate_pdp_ice_parameters(estimator, X, features, response_method,
```

Comment:
Given how similar PDP and ICE are, and how different they are from other (prospectively) things in `inspection`. I would be tempted to implement ICE in the same file as PDP, allowing us to avoid this "pdp_ice" nomenclature.

Reply:
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/154004e41588c4fb9615f408dd4a9cee24b280ca